### ISSUE TITLE
Feature: Allow `perfectionist/sort-classes` to match `property-methods`: properties whose values are methods.

### ISSUE THREAD

**Issue by hugoAlt** [azat-io/eslint-plugin-perfectionist#189]
### What rule do you want to change?

`sort-classes`

### Describe the problem

`eslint/naming-convention` defines `classMethod` as "matches any class method. Also matches properties that have direct function expression or arrow function expression values" (https://typescript-eslint.io/rules/naming-convention/#selector-options).

`perfectionnists/sort-classes` does not recognize arrow methods as class methods.

### Code example

Configuration:

```JSON
"perfectionist/sort-classes": ["error", {
          "groups": [
            "index-signature",
            "static-property",
            "property",
            "private-property",
            [
              "get-method",
              "set-method"
            ],
            "constructor",
            "static-method",
            "private-method",
            "static-private-method",
            "method"
          ]
        }]
```

```typescript
class TestClass {

  attribute: string;

  constructor() {
    // Constructor stuff
  }

  arrowMethod = () => {
    // Arrow method stuff
  }

  method() {
    // Method stuff
  }
}

```

gets reformated into

```typescript
class TestClass {

  arrowMethod = () => {
    // Arrow method stuff
  }

  attribute: string;

  constructor() {
    // Constructor stuff
  }

  method() {
    // Method stuff
  }
}
```

Is there a way to make sure that `arrowMethod` gets grouped with `method`?

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

**Comment by hugop95** [comment]
Feature ready in forked repo (https://github.com/hugop95/eslint-plugin-perfectionist/tree/feat/arrow-method), waiting for merge of https://github.com/azat-io/eslint-plugin-perfectionist/pull/196 before creating PR.

**Comment by OlivierZal** [comment]
Hi @hugop95, I'm not 100% sure "method properties could be recognized as methods and not properties" straight forward.

Indeed arrow-functions misses some properties that class methods have, e.g. they cannot be binded.

Also some IDEs like VSCode don't recognize them as methods:

- either when declared on top of a class (case 1):

<img width="707" alt="Capture d’écran 2024-08-08 à 23 35 57" src="https://github.com/user-attachments/assets/4cc17396-f906-4a57-a446-3f7b992b28b1">

including when this declared property / arrow-function is assigned / used in the class

<img width="698" alt="Capture d’écran 2024-08-08 à 23 36 08" src="https://github.com/user-attachments/assets/7939323e-e7b6-4024-bef0-a71fbdfd4b88">

(note that `this` is not required at assignment)

- or when it's defined / used similarly to a class method (case 2):

<img width="616" alt="Capture d’écran 2024-08-08 à 23 38 01" src="https://github.com/user-attachments/assets/c8be76c3-4728-409b-9d40-4852f616d47b">

except that in this case, ESLint applies a rule which concerns "class methods" and requires `this`)

<img width="698" alt="Capture d’écran 2024-08-08 à 23 36 08" src="https://github.com/user-attachments/assets/15074013-f5a9-45a2-bb2b-bb256328419f">

I believe it’s reasonable to include case 2 in the "default" method group. However, we should also consider creating a specific subgroup (perhaps `arrow-methods`?), as it lacks some properties of class methods, and developers might prefer to categorize them separately.

However, I think that case 1 should be categorized in another separate group (like `call-signature` – to use the term employed by the `[member-ordering](https://typescript-eslint.io/rules/member-ordering/)` rule), since it is a signature and the behavior is a bit different from a class method.

**Comment by OlivierZal** [comment]
Great! I’m not sure if you saw the latest version of my comment, as I rewrote it while you were replying 😊 but I think you got my point.

To summarize my point, case 1 and case 2 differ in that case 1 doesn’t involve any assignment (that's the difference with case 2), which is an important distinction when ordering members.

**Comment by OlivierZal** [comment]
Taking into account the way you want to implement (which sounds great), maybe the title of this issue became misleading.

**Comment by hugop95** [comment]
> Taking into account the way you want to implement (which sounds great), maybe the title of this issue became misleading.

I'll rename it, but I am also reconsidering some things:
what limit should we fix to "official" groups, and when should a feature enter the category of "custom" grouping?

This issue is proposing to add an official group to a case where a property has a specific type of value (a function or arrow function in this case), but what if users want to sort:
- by other value types, at least primitive types (although I don't see a real use case for that yet)
- by the property typing (call signatures, but also any other type, like regrouping `boolean`-typed properties not necessarily assigned together first)

I think that we ultimately need to improve the custom group system to handle cases like that. It should ideally be able to recreate what the official group building does, but in an even more granular way if the user wants, by adding the possibility to filter on decorators (an issue for that exists), or filtering on the property's value (like this issue) or type declaration (your case 1).

TLDR: I am wondering if this issue doesn't fall into the category of custom-grouping due to how specific it is.

**Comment by OlivierZal** [comment]
You raise a good a point, and that's why I wanted to point out a difference between `_private` and `#private` at the first place.

It's not up to me to state as I'm not the design architect of the project. I think that one criteria for `customGroup` is the ability to rely on a regex – that is, when it's not possible to rely on official syntax and something generic like methods, kinds or modifiers, eventually types. 

In my opinion your use case seems sufficiently generic to deserve proper groups, and I don't see how it could possible to make it with even an improved `customGroup` system. 

An enhancement I see for `customGroup` would be the ability to specify the regex with a selector and/or a modifier, e.g. to make a difference between properties and methods starting by `_*` or `on*`, etc.

**Comment by hugop95** [comment]
> You raise a good a point, and that's why I wanted to point out a difference between `_private` and `#private` in the first place.
> 
> It's not up to me to state as I'm not the design architect of the project. I think that one criteria for `customGroup` (which I like) is the ability to rely on a regex – that is, when it's not possible to rely _only_ on official syntax / something generic like methods, kinds or modifiers, eventually types.
> 
> In my opinion your use cases seems sufficiently generic to deserve proper groups: arrow-functions are not exactly methods, and have a clear JS definition and syntax (handled by existing eslint rules). Moreover I don't see how it could possible to handle them even with an improved `customGroup` system (unless by creating generic custom groups 😜).
> 
> An enhancement I see for `customGroup` would be the ability to specify the regex with a selector and/or a modifier, e.g. to make a difference between _properties_ and _methods_ starting by `_*` or `on*`, etc.

Fair enough, I have tried adding that feature to a more advanced custom grouping system here https://github.com/azat-io/eslint-plugin-perfectionist/issues/201, but maybe it would indeed be better to make this an official group.

**Comment by azat-io** [comment]
Added in 41b92d3ed72c47b522b16f6dc919540cd03ff07d.
Released in [v3.2.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v3.2.0).
